architecture-decision-records/0008-keycloak-user-administrators-configuration.md

```diff
@@ -0,0 +1,104 @@
+# 8. Keycloak User Administrators Configuration
+
+**Date:** 2020-06-17
+
+## Context
+
+Keycloak provides a flexible way to organise access for users to its administration functions.
+
+TDR requires user configuration that allows the following:
+* Set of users permission to manage transferring body users, ie the users of the TDR application.
+* Set of users permission to manage keycloak users, ie those internal TNA users who can administer Keycloak itself, including setting up TDR application users.
+
+The configuration needs to consider the following principles:
+* *Least privilege*: only allow the user the minimum level of permissions to perform the requires operations
+* *Elevated privileges*: prevent a user accruing additional permissions above what they require
+* *User data protection*: ensure user data is secure
+
+## Decision
+
+Decided to define user administrators in the "master" realm. 
+
+There is some benefit from having the user administrators in the master realm as such users will not have access to themselves and cannot give themselves elevated privileges.
+
+Having user administrators in the master realm also provides a clear demarcation between transferring body users which will be defined in the TDR realm users. This provides additional protection against inadvertently giving incorrect permissions to transferring body users.
+
+## TDR User Administrators Configuration
+
+User administrators will be users that have permission to manage other users. 
+
+This will include transferring body users and internal TNA users.
+
+"Groups" will be used to define the permissions. This will allow easier administration of the permissions, as the permission will be defined by the group and keycloak users can be added or removed to the groups.
+
+### Options considered
```

Comment:
I wonder if it would be easier to follow this ADR if the options are above the decision? Then in the Decision section you could say that we've chosen option 2.

Reply:
Changed